### PR TITLE
vSphere: Add IPI-specific validation. 

### PR DESCRIPTION
**Pull request by patrickdillon**
vSphere IPI requires fields in the install-config that are optional in UPI, specifically `cluster`, `network`, and the VIPs. In order to not break previous working IPI installs https://github.com/openshift/installer/commit/7e823377edd97177935f08e98fa3bb80a548591d performs this IPI-specific validation when creating tfvars. 

Jira: [CORS-1354](https://issues.redhat.com/browse/CORS-1354)
Related: openshift/release#7989

### CODE REVIEW

**Comment by abhinavdahiya** [comment]
Using ipi UPI in code imo is not great! Let's find more appropriate names

Secondly, I think we should not overload the tfvars asset, rather create a new one that can perform this validation like the permissions check...

**Comment by patrickdillon** [comment]
> Using ipi UPI in code imo is not great! Let's find more appropriate names
> 
> Secondly, I think we should not overload the tfvars asset, rather create a new one that can perform this validation like the permissions check...

@abhinavdahiya PTAL also please check the related PR openshift/release#7989

File: pkg/asset/cluster/tfvars.go [hunk 1]
```diff
@@ -83,6 +83,9 @@ func (t *TerraformVariables) Dependencies() []asset.Asset {
 	return []asset.Asset{
 		&installconfig.ClusterID{},
 		&installconfig.InstallConfig{},
+		// Include PlatformProvisionCheck here but not in Generate
+		// as it performs validation but is not used.
+		&installconfig.PlatformProvisionCheck{},
```
Comment (abhinavdahiya) [inline pkg/asset/cluster/tfvars.go:88]: so we have added this to tfvars and cluster asset, do both of these need this or only one?

File: pkg/asset/installconfig/platformprovisioncheck.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/openshift/installer/pkg/types/gcp"
+	"github.com/openshift/installer/pkg/types/libvirt"
+	"github.com/openshift/installer/pkg/types/none"
+	"github.com/openshift/installer/pkg/types/openstack"
+	"github.com/openshift/installer/pkg/types/ovirt"
+	"github.com/openshift/installer/pkg/types/vsphere"
+)
+
+// PlatformProvisionCheck is an asset that checks platform credentials for the necessary permissions
+// to create a cluster.
```
Comment (abhinavdahiya) [inline pkg/asset/installconfig/platformprovisioncheck.go:20]: this comment doesn't match what this asset does.

File: pkg/asset/installconfig/vsphere/validation.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+	allErrs = append(allErrs, validation.ValidatePlatform(ic.Platform.VSphere, field.NewPath("platform").Child("vsphere"))...)
+
+	return allErrs.ToAggregate()
+}
+
+// ValidateForProvisioning performs platform validation specifically for installer-
+// provisioned infrastructure. In this case, self-hosted networking is a requirement
+// when the installer creates infrastructure for vSphere clusters.
+func ValidateForProvisioning(ic *types.InstallConfig) error {
```
Comment (abhinavdahiya) [inline pkg/asset/installconfig/vsphere/validation.go:26]: shouldn't this call the `Validate` ?

**Comment by abhinavdahiya** [comment]
/approve

overall looks good, just few comments.

File: pkg/asset/installconfig/vsphere/validation.go (hunk 3 above)
Comment (patrickdillon) [reply pkg/asset/installconfig/vsphere/validation.go:26]: I don't think so? `Validate` would already be called when loading installconfig asset in cd4aee1c6ff0c7e0c1ae2a11f0d6888a66efd5ce

File: pkg/asset/cluster/tfvars.go (hunk 1 above)
Comment (patrickdillon) [reply pkg/asset/cluster/tfvars.go]: I removed it from tfvars.. We cache the vsphere image in the tfvars step, but on closer inspection that doesn't need any of this validation.

File: pkg/types/vsphere/validation/platform.go [hunk 4]
```diff
@@ -30,15 +30,49 @@ func ValidatePlatform(p *vsphere.Platform, fldPath *field.Path) field.ErrorList
...
-		if err := validate.IP(p.APIVIP); err != nil {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("apiVIP"), p.APIVIP, err.Error()))
-		}
-		if err := validate.IP(p.IngressVIP); err != nil {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("ingressVIP"), p.IngressVIP, err.Error()))
-		}
-		if err := validate.IP(p.DNSVIP); err != nil {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("dnsVIP"), p.DNSVIP, err.Error()))
-		}
+		allErrs = append(allErrs, validateVIPs(p, fldPath)...)
```
Comment (abhinavdahiya) [inline pkg/types/vsphere/validation/platform.go:33]: wouldn't this also make the vips required in normal platform validation?

Comment (abhinavdahiya) [reply pkg/types/vsphere/validation/platform.go:33]: ah, nvm

**Comment by abhinavdahiya** [comment]
/test e2e-vsphere
/test e2e-vpshere-upi

LGTM, just waiting for these jobs to finish

**Pull request merged**